stackslib/src/chainstate/nakamoto/coordinator/mod.rs

```diff
@@ -53,59 +54,115 @@ use crate::util_lib::db::Error as DBError;
 pub mod tests;
 
 impl<'a, T: BlockEventDispatcher> OnChainRewardSetProvider<'a, T> {
-    pub fn get_reward_set_nakamoto(
+    /// Read a reward_set written while updating .signers
+    /// `debug_log` should be set to true if the reward set loading should
+    ///  log messages as `debug!` instead of `error!` or `info!`. This allows
+    ///  RPC endpoints to expose this without flooding loggers.
+    pub fn read_reward_set_nakamoto(
         &self,
         cycle_start_burn_height: u64,
         chainstate: &mut StacksChainState,
         burnchain: &Burnchain,
         sortdb: &SortitionDB,
         block_id: &StacksBlockId,
+        debug_log: bool,
     ) -> Result<RewardSet, Error> {
-        // TODO: this method should read the .signers contract to get the reward set entries.
-        //   they will have been set via `NakamotoChainState::check_and_handle_prepare_phase_start()`.
         let cycle = burnchain
             .block_height_to_reward_cycle(cycle_start_burn_height)
             .expect("FATAL: no reward cycle for burn height");
+        // figure out the block ID
+        let Some(coinbase_height_of_calculation) = chainstate
+            .eval_boot_code_read_only(
+                sortdb,
+                block_id,
+                SIGNERS_NAME,
+                &format!("(map-get? cycle-set-height u{})", cycle),
+            )?
+            .expect_optional()
+            .map(|x| u64::try_from(x.expect_u128()).expect("FATAL: block height exceeded u64"))
+        else {
+            if debug_log {
```

Comment:
Is there a way to DRY this up?  Perhaps a custom logging function?

Reply:
Yeah, a new macro definition would be necessary. I can add.

Reply:
Added a local macro for this.